app/services/courses_query.rb

```diff
@@ -22,6 +23,33 @@ def initialize(params:)
   end
 
   def call
-    @scope.distinct
+    @scope = visa_sponsorship_scope
+    @scope = @scope.distinct
+
+    log_query_info
+
+    @scope
+  end
+
+  def visa_sponsorship_scope
+    return @scope if params[:can_sponsor_visa].blank?
+
+    @applied_scopes[:can_sponsor_visa] = params[:can_sponsor_visa]
+
+    @scope
+      .where(
+        can_sponsor_student_visa: true
+      )
+      .or(
+        @scope.where(
+          can_sponsor_skilled_worker_visa: true
+        )
+      )
+  end
+
+  private
+
+  def log_query_info
+    CoursesQuery::Logger.new(@applied_scopes, @scope).call
```

Comment:
I wonder if passing the query object itself makes sense?

```suggestion
    CoursesQuery::Logger.new(self).call
```

Reply:
Done